src/ir/var.rs

```diff
@@ -130,6 +130,60 @@ fn default_macro_constant_type(value: i64) -> IntKind {
     }
 }
 
+/// Determines whether a cursor that is pointing into a CXCursor_MacroDefinition
+/// is pointing into a function-like macro. If so, calls the func_macro callback
+/// and returns `Err(ParseError::Continue)` to signal to skip further
+/// processing.  If version to UTF-8 fails (it is performed only where it should
+/// be infallible), then `Err(ParseError::Continue)` is returned as well.
+fn handle_function_macro(
+    cursor: &clang::Cursor,
+    func_macro: impl FnOnce(&str, &[&[u8]]),
+) -> Result<(), ParseError> {
+    // cexpr explicitly does not handle function-like macros, except for a
+    // degenerate case in which it behaves in a non-conformant way. We prevent
+    // cexpr from ever seeing function-like macros by checking for a parenthesis
+    // token immediately adjacent to (that is, abutting) the first token in the
+    // macro definition.
+    let tokens: Vec<_> = cursor.tokens().iter().collect();
+    match tokens.get(0..2) {
+        Some([first, second]) if first.is_abutting(&second) => {
+            // To be doubly sure we are in a function macro definition, we could
+            // check the second macro's type and content, but in order for the
+            // first two macros after `#define` to be adjacent with no
+            // intervening space, the second one has to be an open parenthesis
+            // anyway, so the fact that this code is executing at all implies
+            // that we are defining a function macro.
+            let joined: Vec<_> = tokens
+                .into_iter()
+                .map(|t| t.spelling().to_owned())
+                .collect();
+            let boundary = joined
+                .iter()
+                .position(|s| s == b")")
+                .map(|n| n + 1) // skip past ')'
+                .unwrap_or(joined.len());
+
+            let (left, right) = joined.split_at(boundary);
+            let left = String::from_utf8(left.concat())
+                .map_err(|_| ParseError::Continue)?;
+            let right: Vec<_> = right.iter().map(Vec::as_slice).collect();
+            // Drop last token with LLVM < 4.0, due to an LLVM bug.
+            //
+            // See:
+            //   https://bugs.llvm.org//show_bug.cgi?id=9069
+            let right = match (right.len(), crate::clang_version().parsed) {
+                (len, Some((v, _))) if len > 0 && v < 4 => &right[..len - 1],
+                _ => &right[..],
+            };
+            func_macro(&left, right);
+
+            // We handled the macro, skip future macro processing.
```

Comment:
It is a good question. I do not know what future "proper" support for function-like macros should look like, but I do not think this implementation detail precludes future proper support.

It is *almost* true that the only user-visible change from the current PR is the addition of the callback, but because `cexpr` will interpret some simple function-like macros as if they were integer-defining macros, too, both `func_macro` and `int_macro` (I think a similar thing is true for `str_macro`) would be called for the same macro in the following case if we did not bail out of `parse` early :
```c
#define TESTMACRO_INTEGER 42
#define TESTMACRO_FUNCTIONAL_EMPTY(TESTMACRO_INTEGER)
```
so technically another user-visible change is that some function-like macros that were previously invalidly parsed will no longer be invalidly parsed. I hope that is a good thing.

Thus I see no alternative to bailing out early once we identify a function macro.